File: src-tauri/src/database/article.rs
use std::time::{SystemTime, UNIX_EPOCH};
use rusqlite::{Connection, params, Result, Row};
use serde::{Serialize, Deserialize};
use crate::database::operation::insert_operation;

#[derive(Serialize, Deserialize, Debug)]
pub struct Article {
    pub id: i64,
    pub create_time: i64,
    pub update_time: i64,
    pub title: String,
    pub author: String,
    pub tags: Vec<String>,
    pub links: Vec<i64>,
    pub content: String,
    pub banner_bg: String,
    pub is_top: bool,
    pub is_delete: bool,
}

fn get_article_from_query_result(row: &Row) -> Article {
    let tags: String = row.get(5).unwrap();
    let tags: Vec<String> = serde_json::from_str(&tags).unwrap();
    let links: String = row.get(6).unwrap();
    let links: Vec<i64> = serde_json::from_str(&links).unwrap();
    Article {
        id: row.get(0).unwrap(),
        create_time: row.get(1).unwrap(),
        update_time: row.get(2).unwrap(),
        title: row.get(3).unwrap(),
        author: row.get(4).unwrap(),
        tags,
        links,
        content: row.get(7).unwrap(),
        banner_bg: row.get(8).unwrap(),
        is_top: row.get(9).unwrap(),
        is_delete: row.get(10).unwrap(),
    }
}


pub fn init_article_table(conn: &Connection) -> Result<()> {
    conn.execute(
        "CREATE TABLE IF NOT EXISTS articles (
            id INTEGER PRIMARY KEY,
            title TEXT NOT NULL,
            author TEXT,
            create_time INTEGER NOT NULL,
            update_time INTEGER NOT NULL,
            tags TEXT,
            links TEXT,
            content TEXT,
            banner_bg TEXT,
            is_top INTEGER DEFAULT 0,
            is_delete INTEGER DEFAULT 0
        )",
        [],
    )?;
    Ok(())
}

pub fn upgrade_article_table(conn: &Connection, old_version: i64, new_version: i64) -> Result<()> {
    if old_version == new_version {
        return Ok(());
    }
    if old_version < 1 {
        // 如果 articles 表没有 links 字段，添加 links 字段
        let mut stmt = conn.prepare("SELECT sql FROM sqlite_master WHERE type = 'table' AND name = 'articles'")?;
        let mut rows = stmt.query([])?;
        if let Some(row) = rows.next()? {
            let sql: String = row.get(0)?;
            println!("sql: {}", sql);
            if !sql.contains("links") {
                conn.execute(
                    "ALTER TABLE articles ADD COLUMN links TEXT",
                    [],
                )?;
            }
        }
    }
    if old_version < 2 {
        // 如果 articles 表没有 banner_bg 和 is_top 字段，添加 banner_bg 和 is_top 字段
        let mut stmt = conn.prepare("SELECT sql FROM sqlite_master WHERE type = 'table' AND name = 'articles'")?;
        let mut rows = stmt.query([])?;
        if let Some(row) = rows.next()? {
            let sql: String = row.get(0)?;
            if !sql.contains("banner_bg") {
                conn.execute(
                    "ALTER TABLE articles ADD COLUMN banner_bg TEXT DEFAULT ''",
                    [],
                )?;
            }
            if !sql.contains("is_top") {
                conn.execute(
                    "ALTER TABLE articles ADD COLUMN is_top INTEGER DEFAULT 0",
                    [],
                )?;
            }
            if !sql.contains("is_delete") {
                conn.execute(
                    "ALTER TABLE articles ADD COLUMN is_delete INTEGER DEFAULT 0",
                    [],
                )?;
            }
        }
    }
    Ok(())
}

pub fn create_article(
    conn: &Connection, 
    title: String, 
    author: String, 
    tags: Vec<String>, 
    links: Vec<i64>, 
    content: String, 
    banner_bg: String, 
    is_top: bool
) -> Result<Article> {
    let now = SystemTime::now().duration_since(UNIX_EPOCH).unwrap().as_millis() as i64;
    let mut stmt = conn.prepare("INSERT INTO articles (title, author, create_time, update_time, tags, links, content, banner_bg, is_top) VALUES (?1, ?2, ?3, ?4, ?5, ?6, ?7, ?8, ?9)")?;
    let tags_str = serde_json::to_string(&tags).unwrap();
    let links_str = serde_json::to_string(&links).unwrap();
    let res = stmt.insert(params![title, author, now, now, tags_str, links_str, content, banner_bg, is_top])?;

    insert_operation(conn, res, "article".to_string(), "insert".to_string())?;

    // 根据 id 查询插入的文章，返回
    let mut stmt = conn.prepare("SELECT id, create_time, update_time, title, author, tags, links, content, banner_bg, is_top, is_delete FROM articles WHERE id = ?1")?;
    let mut rows = stmt.query(params![res])?;
    let row = rows.next()?.unwrap();
    Ok(get_article_from_query_result(&row))
}

pub fn update_article(
    conn: &Connection, id: i64,
    title: String,
    author: String,
    tags: Vec<String>,
    links: Vec<i64>,
    content: String,
    banner_bg: String,
    is_top: bool,
) -> Result<Article> {
    let now = SystemTime::now().duration_since(UNIX_EPOCH).unwrap().as_millis() as i64;
    let mut stmt = conn.prepare("UPDATE articles SET title = ?1, author = ?2, update_time = ?3, tags = ?4, links = ?5, content = ?6, banner_bg = ?7, is_top = ?8 WHERE id = ?9")?;
    let tags_str = serde_json::to_string(&tags).unwrap();
    let links_str = serde_json::to_string(&links).unwrap();
    stmt.execute(params![title, author, now, tags_str, links_str, content, banner_bg, is_top, id])?;

    insert_operation(conn, id, "article".to_string(), "update".to_string())?;

    // 从 document_items 表中找到 is_article 为 1 且 article 为 id 的所有记录，并更新其 content 和 update_time 和 title
    let mut stmt = conn.prepare("UPDATE document_items SET update_time = ?1, content = ?2, title = ?3 WHERE is_article = 1 AND article_id = ?4")?;
    stmt.execute(params![now as i64, content, title, id])?;

    // 查询刚刚更新的文章，然后返回
    let mut stmt = conn.prepare("SELECT id, create_time, update_time, title, author, tags, links, content, banner_bg, is_top, is_delete FROM articles WHERE id = ?1")?;
    let mut rows = stmt.query(params![id])?;
    let row = rows.next()?.unwrap();
    Ok(get_article_from_query_result(&row))
}

pub fn find_article(conn: &Connection, id: i64) -> Result<Article> {
    let mut stmt = conn.prepare("SELECT id, create_time, update_time, title, author, tags, links, content, banner_bg, is_top, is_delete FROM articles WHERE id = ?1")?;
    let mut rows = stmt.query(params![id])?;
    let row = rows.next()?.unwrap();
    Ok(get_article_from_query_result(&row))
}

pub fn find_all_articles(conn: &Connection) -> Result<Vec<Article>> {
    // 根据更新时间倒序
    let sql = "SELECT id, create_time, update_time, title, author, tags, links, content, banner_bg, is_top, is_delete FROM articles WHERE is_delete = 0 ORDER BY update_time DESC";
    let mut stmt = conn.prepare(sql)?;
    let mut rows = stmt.query([])?;
    let mut res = Vec::new();
    while let Some(row) = rows.next()? {
        res.push(get_article_from_query_result(&row));
    }
    Ok(res)
}

pub fn delete_article(conn: &Connection, id: i64) -> Result<usize> {
    let mut stmt = conn.prepare("UPDATE articles SET is_delete = 1 WHERE id = ?1")?;
    let res = stmt.execute(params![id])?;
    Ok(res)
}

pub fn update_banner_bg(conn: &Connection, id: i64, banner_bg: String) -> Result<usize> {
    let now = SystemTime::now().duration_since(UNIX_EPOCH).unwrap().as_millis() as i64;
    let mut stmt = conn.prepare("UPDATE articles SET banner_bg = ?1, update_time = ?2 WHERE id = ?3")?;
    let res = stmt.execute(params![banner_bg, now, id])?;
    Ok(res)
}

pub fn update_is_top(conn: &Connection, id: i64, is_top: bool) -> Result<usize> {
    let now = SystemTime::now().duration_since(UNIX_EPOCH).unwrap().as_millis() as i64;
    let mut stmt = conn.prepare("UPDATE articles SET is_top = ?1, update_time = ?2 WHERE id = ?3")?;
    let res = stmt.execute(params![is_top as i32, now, id])?;
    Ok(res)
}

File: src-electron/main/modules/tables/article.ts
import Database from "better-sqlite3";
import { ICreateArticle, IUpdateArticle, IArticle } from "@/types";
import Operation from "./operation";
import { getMarkdown } from "@/utils/markdown.ts";
import { basename } from "node:path";
import { BrowserWindow } from "electron";
import FTSTable from "./fts";
import VecDocumentTable from "./vec-document";
import ContentTable from "./content";
import ProjectTable from "./project";
import log from "electron-log";

export default class ArticleTable {
  static initTable(db: Database.Database) {
    const createTableSql = `
      CREATE TABLE IF NOT EXISTS articles (
        id INTEGER PRIMARY KEY,
        title TEXT NOT NULL,
        author TEXT,
        create_time INTEGER NOT NULL,
        update_time INTEGER NOT NULL,
        tags TEXT,
        links TEXT,
        content_id INTEGER,
        banner_bg TEXT,
        banner_position TEXT DEFAULT 'center',
        is_top INTEGER DEFAULT 0,
        is_delete INTEGER DEFAULT 0,
        FOREIGN KEY(content_id) REFERENCES contents(id)
      )`;
    db.exec(createTableSql);
  }

  static upgradeTable(db: Database.Database) {
    // 为所有文章添加 FTS 索引
    log.info("开始为所有文章添加 FTS 索引");
    const articles = this.getAllArticles(db);
    for (const article of articles) {
      if (!article.content || !article.content.length) continue;

      // 检查是否已有索引或索引是否过期
      const indexInfo = FTSTable.checkIndexExists(db, article.id, "article");

      // 如果索引不存在或已过期，则添加/更新索引
      if (!indexInfo || indexInfo.updateTime < article.update_time) {
        FTSTable.indexContent(db, {
          id: article.id,
          content: getMarkdown(article.content),
          type: "article",
          updateTime: article.update_time,
        });
        log.info(`已为文章 ${article.id} 添加/更新 FTS 索引`);
      }
    }
  }

  static getListenEvents() {
    return {
      "create-article": this.createArticle.bind(this),
      "create-article-from-project-item":
        this.createArticleFromProjectItem.bind(this),
      "update-article": this.updateArticle.bind(this),
      "delete-article": this.deleteArticleById.bind(this),
      "get-all-articles": this.getAllArticles.bind(this),
      "get-article-by-id": this.getArticleById.bind(this),
      "update-article-is-top": this.updateArticleIsTop.bind(this),
      "update-article-banner-bg": this.updateArticleBannerBg.bind(this),
      "update-article-banner-position":
        this.updateArticleBannerPosition.bind(this),
    };
  }

  static parseArticle(article: any): IArticle {
    let content = [];
    let count = 0;

    // 直接使用JOIN查询结果中的内容
    if (article.content) {
      try {
        content = JSON.parse(article.content);
        count = article.count || 0;
      } catch (error) {
        console.error("Error parsing content:", error);
      }
    }

    // 使用文章和内容表中最大的更新时间
    const updateTime = article.content_update_time
      ? Math.max(article.update_time, article.content_update_time)
      : article.update_time;

    return {
      id: article.id,
      create_time: article.create_time,
      update_time: updateTime,
      tags: JSON.parse(article.tags),
      title: article.title,
      content: content,
      author: article.author || "",
      links: JSON.parse(article.links || "[]"),
      bannerBg: article.banner_bg || "",
      bannerPosition: article.banner_position || "center",
      isTop: article.is_top || false,
      isDelete: article.is_delete || false,
      count: count,
      contentId: article.content_id,
    };
  }

  static getAllArticles(db: Database.Database) {
    // 按照创建时间倒序
    const stmt = db.prepare(`
      SELECT a.id, a.create_time, a.update_time, a.tags, a.title, a.author, a.links,
             a.banner_bg, a.banner_position, a.is_top, a.is_delete, a.content_id,
             c.content, c.count, c.update_time as content_update_time
      FROM articles a
      LEFT JOIN contents c ON a.content_id = c.id
      WHERE a.is_delete = 0 
      ORDER BY a.create_time DESC
    `);
    const articles = stmt.all();
    return articles.map((article) => this.parseArticle(article));
  }

  static getArticleById(
    db: Database.Database,
    articleId: number | bigint,
  ): IArticle {
    const stmt = db.prepare(`
      SELECT a.id, a.create_time, a.update_time, a.tags, a.title, a.author, a.links,
             a.banner_bg, a.banner_position, a.is_top, a.is_delete, a.content_id,
             c.content, c.count, c.update_time as content_update_time
      FROM articles a
      LEFT JOIN contents c ON a.content_id = c.id
      WHERE a.id = ? AND a.is_delete = 0
    `);
    const article = stmt.get(articleId);
    return this.parseArticle(article);
  }

  static getArticleByIds(
    db: Database.Database,
    articleIds: number[],
  ): IArticle[] {
    if (articleIds.length === 0) return [];

    const placeholders = articleIds.map(() => "?").join(",");
    const stmt = db.prepare(`
      SELECT a.id, a.create_time, a.update_time, a.tags, a.title, a.author, a.links,
             a.banner_bg, a.banner_position, a.is_top, a.is_delete, a.content_id,
             c.content, c.count, c.update_time as content_update_time
      FROM articles a
      LEFT JOIN contents c ON a.content_id = c.id
      WHERE a.id IN (${placeholders}) AND a.is_delete = 0
    `);
    const articles = stmt.all(articleIds);
    return articles.map((article) => this.parseArticle(article));
  }

  static createArticle(
    db: Database.Database,
    article: ICreateArticle,
  ): IArticle {
    const {
      tags,
      title,
      content,
      bannerBg,
      bannerPosition,
      isDelete,
      isTop,
      count,
    } = article;

    // 如果没有提供contentId，则创建content记录
    const contentId = ContentTable.createContent(db, {
      content: content,
      count: count,
    });

    // 创建article记录
    const stmt = db.prepare(
      "INSERT INTO articles (create_time, update_time, tags, title, content_id, banner_bg, banner_position, is_delete, is_top) VALUES (?, ?, ?, ?, ?, ?, ?, ?, ?)",
    );
    const now = Date.now();
    const res = stmt.run(
      now,
      now,
      JSON.stringify(tags),
      title,
      contentId,
      bannerBg,
      bannerPosition || "center",
      Number(isDelete),
      Number(isTop),
    );
    const createdArticleId = res.lastInsertRowid;

    Operation.insertOperation(db, "article", "insert", createdArticleId, now);

    return this.getArticleById(db, createdArticleId);
  }

  static createArticleFromProjectItem(
    db: Database.Database,
    projectItemId: number,
  ): IArticle | null {
    const projectItem = ProjectTable.getProjectItem(db, projectItemId);
    if (!projectItem) {
      return null;
    }
    const contentId = projectItem.contentId;
    const title = projectItem.title;

    ContentTable.incrementRefCount(db, contentId);

    const stmt = db.prepare(
      "INSERT INTO articles (create_time, update_time, tags, title, content_id, banner_bg, banner_position, is_delete, is_top) VALUES (?, ?, ?, ?, ?, ?, ?, ?, ?)",
    );
    const now = Date.now();
    const res = stmt.run(
      now,
      now,
      JSON.stringify([]),
      title,
      contentId,
      "",
      "center",
      0,
      0,
    );
    const createdArticleId = res.lastInsertRowid;

    Operation.insertOperation(db, "article", "insert", createdArticleId, now);

    return this.getArticleById(db, createdArticleId);
  }

  static updateArticle(
    db: Database.Database,
    article: IUpdateArticle,
    ...res: any[]
  ): IArticle {
    const {
      tags,
      title,
      id,
      bannerBg,
      bannerPosition,
      isTop,
      isDelete,
      contentId,
    } = article;
    const win = res[res.length - 1];
    const now = Date.now();

    // 获取当前的article信息
    const currentArticle = this.getArticleById(db, id);

    // 更新article记录
    const stmt = db.prepare(
      "UPDATE articles SET update_time = ?, tags = ?, title = ?, content_id = ?, banner_bg = ?, banner_position = ?, is_top = ?, is_delete = ? WHERE id = ?",
    );
    stmt.run(
      now,
      JSON.stringify(tags || currentArticle.tags),
      title || currentArticle.title,
      contentId,
      bannerBg || currentArticle.bannerBg,
      bannerPosition || currentArticle.bannerPosition,
      Number(isTop ?? currentArticle.isTop),
      Number(isDelete ?? currentArticle.isDelete),
      id,
    );

    Operation.insertOperation(db, "article", "update", id, now);

    BrowserWindow.getAllWindows().forEach((window) => {
      if (window !== win && !window.isDestroyed()) {
        window.webContents.send("article:updated", {
          databaseName: basename(db.name),
          articleId: id,
        });
      }
    });

    return this.getArticleById(db, id);
  }

  static deleteArticleById(db: Database.Database, articleId: number): number {
    // 获取文章信息，以获取contentId
    const articleInfo = this.getArticleById(db, articleId);

    const stmt = db.prepare("DELETE FROM articles WHERE id = ?");
    const result = stmt.run(articleId);

    if (result.changes > 0) {
      FTSTable.removeIndexByIdAndType(db, articleId, "article");
      VecDocumentTable.removeIndexByIdAndType(db, articleId, "article");
      Operation.insertOperation(db, "article", "delete", articleId, Date.now());

      const documentItemStmt = db.prepare(
        "UPDATE document_items SET is_article = 0 WHERE is_article = 1 AND article_id = ?",
      );
      documentItemStmt.run(articleId);

      const projectItemStmt = db.prepare(
        "UPDATE project_item SET ref_type = '' WHERE ref_type = 'article' AND ref_id = ?",
      );
      projectItemStmt.run(articleId);

      if (articleInfo && articleInfo.contentId) {
        // 删除关联的content记录（减少引用计数）
        ContentTable.deleteContent(db, articleInfo.contentId);
      }
    }

    return result.changes;
  }

  static updateArticleIsTop(
    db: Database.Database,
    id: number,
    isTop: boolean,
    ...res: any[]
  ): IArticle {
    const win = res[res.length - 1];
    const stmt = db.prepare("UPDATE articles SET is_top = ? WHERE id = ?");
    stmt.run(Number(isTop), id);

    BrowserWindow.getAllWindows().forEach((window) => {
      if (window !== win && !window.isDestroyed()) {
        window.webContents.send("article:updated", {
          databaseName: basename(db.name),
          articleId: id,
        });
      }
    });

    return this.getArticleById(db, id);
  }

  static updateArticleBannerBg(
    db: Database.Database,
    id: number,
    bannerBg: string,
    ...res: any[]
  ): IArticle {
    const win = res[res.length - 1];
    const stmt = db.prepare("UPDATE articles SET banner_bg = ? WHERE id = ?");
    stmt.run(bannerBg, id);

    BrowserWindow.getAllWindows().forEach((window) => {
      if (window !== win && !window.isDestroyed()) {
        window.webContents.send("article:updated", {
          databaseName: basename(db.name),
          articleId: id,
        });
      }
    });

    return this.getArticleById(db, id);
  }

  static updateArticleBannerPosition(
    db: Database.Database,
    id: number,
    bannerPosition: string,
    ...res: any[]
  ): IArticle {
    const win = res[res.length - 1];
    const stmt = db.prepare(
      "UPDATE articles SET banner_position = ? WHERE id = ?",
    );
    stmt.run(bannerPosition, id);

    BrowserWindow.getAllWindows().forEach((window) => {
      if (window !== win && !window.isDestroyed()) {
        window.webContents.send("article:updated", {
          databaseName: basename(db.name),
          articleId: id,
        });
      }
    });

    return this.getArticleById(db, id);
  }
}
